Cannot paste

### What happened?

Everything is working fine but when I try to send a pasted image "/paste" it does not send.

### Steps to reproduce

run siggy.
copy an image (screenshot to clipboard in this case).
type "/paste" and return
message fails to send.

### Operating system

macOS

### Terminal emulator

kitty

### siggy version

v1.4.1

### Logs or screenshots

```
[20:08:15.765] === siggy debug session started (PII redacted) ===
[20:08:22.613] event: ContactList(count=36)
[20:08:22.634] event: GroupList(count=17)
[20:08:22.661] event: TypingIndicator(from=+4***...858, typing=true)
[20:08:22.674] event: ReceiptReceived(READ from=+4***...888, count=5)
[20:08:22.681] event: ReadSyncReceived(count=1)
[20:08:22.683] event: TypingIndicator(from=+4***...858, typing=true)
[20:08:22.704] event: IdentityList(count=49)
[20:08:22.705] event: MessageReceived(from=+3***...785, body=[msg: 6 chars], attachments=0, group=true)
[20:08:22.712] event: ReceiptReceived(DELIVERY from=+4***...910, count=1)
[20:08:22.713] receipt: READ from +4***...888 -> Read
[20:08:22.714] receipt: DELIVERY from +4***...910 -> Delivered
[20:08:22.714] event: ReceiptReceived(DELIVERY from=+4***...528, count=1)
[20:08:22.716] event: ReceiptReceived(DELIVERY from=+4***...858, count=1)
[20:08:22.723] event: MessageReceived(from=+4***...858, body=[msg: 53 chars], attachments=0, group=true)
[20:08:22.725] event: ReceiptReceived(READ from=+4***...858, count=1)
[20:08:22.727] event: ReadSyncReceived(count=1)
[20:08:22.765] receipt: DELIVERY from +4***...528 -> Delivered
[20:08:22.765] receipt: DELIVERY from +4***...858 -> Delivered
[20:08:22.765] receipt: READ from +4***...858 -> Read
[20:08:26.330] event: TypingIndicator(from=+4***...858, typing=true)
[20:08:31.842] event: TypingIndicator(from=+4***...858, typing=true)
[20:08:34.114] event: TypingIndicator(from=+4***...858, typing=true)
[20:08:38.842] event: TypingIndicator(from=+4***...858, typing=true)
[20:08:40.924] event: MessageReceived(from=+4***...858, body=[msg: 10 chars], attachments=0, group=true)
[20:08:41.446] send: to=IuNs...dvQ= ts=1773428921445
[20:08:41.993] signal-cli stderr: ERROR SignalJsonRpcDispatcherHandler - Command execution failed with unexpected error
[20:08:41.993] signal-cli stderr: org.asamk.signal.commands.exceptions.UnexpectedErrorException: Failed to send message: /var/folders/tw/fb5yl9js3pl2675x572n9qxc0000gn/T/siggy-paste-16808/clipboard_20260313_200838.985.png: /var/folders/tw/fb5yl9js3pl2675x572n9qxc0000gn/T/siggy-paste-16808/clipboard_20260313_200838.985.png (No such file or directory) (AttachmentInvalidException)
[20:08:41.993] signal-cli stderr: 	at org.asamk.signal.commands.SendCommand.handleCommand(SendCommand.java:268)
[20:08:41.993] signal-cli stderr: 	at org.asamk.signal.commands.JsonRpcLocalCommand.handleCommand(JsonRpcLocalCommand.java:24)
[20:08:41.993] signal-cli stderr: 	at org.asamk.signal.commands.JsonRpcLocalCommand.handleCommand(JsonRpcLocalCommand.java:16)
[20:08:41.993] signal-cli stderr: 	at org.asamk.signal.jsonrpc.SignalJsonRpcCommandHandler$CommandRunnerImpl.handleCommand(SignalJsonRpcCommandHandler.java:180)
[20:08:41.993] signal-cli stderr: 	at org.asamk.signal.jsonrpc.SignalJsonRpcCommandHandler.parseParamsAndRunCommand(SignalJsonRpcCommandHandler.java:303)
[20:08:41.993] signal-cli stderr: 	at org.asamk.signal.jsonrpc.SignalJsonRpcCommandHandler.runCommand(SignalJsonRpcCommandHandler.java:241)
[20:08:41.994] signal-cli stderr: 	at org.asamk.signal.jsonrpc.SignalJsonRpcCommandHandler.handleRequest(SignalJsonRpcCommandHandler.java:101)
[20:08:41.994] signal-cli stderr: 	at org.asamk.signal.jsonrpc.SignalJsonRpcDispatcherHandler.lambda$handleConnection$1(SignalJsonRpcDispatcherHandler.java:139)
[20:08:41.994] signal-cli stderr: 	at org.asamk.signal.jsonrpc.JsonRpcReader.handleRequest(JsonRpcReader.java:134)
[20:08:41.994] signal-cli stderr: 	at org.asamk.signal.jsonrpc.JsonRpcReader.handleMessage(JsonRpcReader.java:85)
[20:08:41.994] signal-cli stderr: 	at org.asamk.signal.jsonrpc.JsonRpcReader.lambda$readMessages$0(JsonRpcReader.java:72)
[20:08:41.994] signal-cli stderr: 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:545)
[20:08:41.994] signal-cli stderr: 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:328)
[20:08:41.994] signal-cli stderr: 	at java.base/java.util.concurrent.ThreadPerTaskExecutor$ThreadBoundFuture.run(ThreadPerTaskExecutor.java:323)
[20:08:41.994] signal-cli stderr: 	at java.base/java.lang.VirtualThread.run(VirtualThread.java:456)
[20:08:41.994] signal-cli stderr: Caused by: org.asamk.signal.manager.api.AttachmentInvalidException: /var/folders/tw/fb5yl9js3pl2675x572n9qxc0000gn/T/siggy-paste-16808/clipboard_20260313_200838.985.png: /var/folders/tw/fb5yl9js3pl2675x572n9qxc0000gn/T/siggy-paste-16808/clipboard_20260313_200838.985.png (No such file or directory)
[20:08:41.994] signal-cli stderr: 	at org.asamk.signal.manager.util.AttachmentUtils.createAttachmentStream(AttachmentUtils.java:24)
[20:08:41.994] signal-cli stderr: 	at org.asamk.signal.manager.helper.AttachmentHelper.createAttachmentStreams(AttachmentHelper.java:71)
[20:08:41.994] signal-cli stderr: 	at org.asamk.signal.manager.helper.AttachmentHelper.uploadAttachments(AttachmentHelper.java:48)
[20:08:41.994] signal-cli stderr: 	at org.asamk.signal.manager.internal.ManagerImpl.applyMessage(ManagerImpl.java:844)
[20:08:41.994] signal-cli stderr: 	at org.asamk.signal.manager.internal.ManagerImpl.sendMessage(ManagerImpl.java:802)
[20:08:41.994] signal-cli stderr: 	at org.asamk.signal.commands.SendCommand.handleCommand(SendCommand.java:259)
[20:08:41.994] signal-cli stderr: 	... 14 common frames omitted
[20:08:41.995] rpc error: method=send error=JsonRpcError { code: -32603, message: "Failed to send message: /var/folders/tw/fb5yl9js3pl2675x572n9qxc0000gn/T/siggy-paste-16808/clipboard_20260313_200838.985.png: /var/folders/tw/fb5yl9js3pl2675x572n9qxc0000gn/T/siggy-paste-16808/clipboard_20260313_200838.985.png (No such file or directory) (AttachmentInvalidException) (UnexpectedErrorException)" }
[20:08:41.995] event: SendFailed(rpc=5ef8aa8f-e585-4cea-a5ce-b1bdeaf9d20a)
[20:08:43.640] event: ReceiptReceived(DELIVERY from=+4***...410, count=1)
[20:08:43.644] receipt: DELIVERY from +4***...410 -> Delivered
[20:08:43.833] event: ReceiptReceived(DELIVERY from=+4***...410, count=1)
[20:08:43.851] receipt: DELIVERY from +4***...410 -> Delivered
[20:09:01.211] event: TypingIndicator(from=+4***...858, typing=true)
[20:09:02.573] event: TypingIndicator(from=+4***...858, typing=true)
[20:09:06.623] event: MessageReceived(from=+4***...858, body=[msg: 32 chars], attachments=0, group=true)
[20:09:07.060] event: ReadSyncReceived(count=1)
[20:09:15.458] event: MessageReceived(from=+3***...785, body=[msg: 20 chars], attachments=1, group=true)
[20:09:15.461] event: ReceiptReceived(DELIVERY from=+4***...858, count=1)
[20:09:15.483] receipt: DELIVERY from +4***...858 -> Delivered
[20:09:15.556] event: ReceiptReceived(READ from=+4***...858, count=1)
[20:09:15.565] event: ReactionReceived(conv=IuNs...dvQ=, from=+4***...910, emoji=😂, target_ts=1773428946717, remove=false)
[20:09:15.591] receipt: READ from +4***...858 -> Read
[20:09:15.670] event: ReceiptReceived(READ from=+4***...910, count=2)
[20:09:15.697] receipt: READ from +4***...910 -> Read
[20:09:15.769] event: ReadSyncReceived(count=1)
[20:09:16.194] event: ReceiptReceived(DELIVERY from=+4***...410, count=1)
[20:09:16.218] receipt: DELIVERY from +4***...410 -> Delivered
[20:09:17.058] event: ReceiptReceived(DELIVERY from=+4***...528, count=1)
[20:09:17.091] receipt: DELIVERY from +4***...528 -> Delivered
[20:09:20.877] event: ReceiptReceived(DELIVERY from=+4***...910, count=1)
[20:09:20.921] receipt: DELIVERY from +4***...910 -> Delivered
[20:09:20.984] event: ReceiptReceived(READ from=+4***...910, count=1)
[20:09:21.031] receipt: READ from +4***...910 -> Read
[20:09:26.646] event: TypingIndicator(from=+4***...910, typing=true)
[20:09:27.365] event: ReactionReceived(conv=IuNs...dvQ=, from=+3***...785, emoji=😂, target_ts=1773428946717, remove=false)
[20:09:27.459] event: ReceiptReceived(DELIVERY from=+4***...528, count=1)
[20:09:27.465] event: ReceiptReceived(DELIVERY from=+4***...858, count=1)
[20:09:27.468] receipt: buffering DELIVERY from +4***...528 (no matching ts)
[20:09:27.468] receipt: buffering DELIVERY from +4***...858 (no matching ts)
[20:09:27.471] event: ReceiptReceived(DELIVERY from=+4***...910, count=1)
[20:09:27.524] receipt: buffering DELIVERY from +4***...910 (no matching ts)
[20:09:27.570] event: ReceiptReceived(DELIVERY from=+4***...410, count=1)
[20:09:27.580] receipt: buffering DELIVERY from +4***...410 (no matching ts)
[20:09:36.525] event: MessageReceived(from=+4***...910, body=[msg: 37 chars], attachments=0, group=true)
[20:09:38.559] event: ReceiptReceived(DELIVERY from=+4***...747, count=1)
[20:09:38.583] receipt: DELIVERY from +4***...747 -> Delivered
[20:09:38.756] event: ReceiptReceived(DELIVERY from=+4***...747, count=1)
[20:09:38.790] receipt: DELIVERY from +4***...747 -> Delivered
[20:09:38.882] event: ReceiptReceived(DELIVERY from=+4***...747, count=1)
[20:09:38.898] receipt: buffering DELIVERY from +4***...747 (no matching ts)
[20:09:46.210] event: TypingIndicator(from=+4***...858, typing=true)
[20:09:49.989] event: TypingIndicator(from=+4***...910, typing=true)
[20:09:54.256] event: MessageReceived(from=+4***...910, body=[msg: 15 chars], attachments=0, group=true)
[20:09:54.687] event: TypingIndicator(from=+4***...858, typing=true)
[20:09:59.577] event: TypingIndicator(from=+4***...858, typing=true)
[20:10:00.988] event: MessageReceived(from=+4***...858, body=[msg: 48 chars], attachments=0, group=true)
[20:10:02.894] event: TypingIndicator(from=+4***...910, typing=true)
[20:10:07.817] event: MessageReceived(from=+4***...910, body=[msg: 10 chars], attachments=0, group=true)
[20:10:15.452] event: MessageReceived(from=+3***...785, body=[msg: 46 chars], attachments=1, group=true)
[20:10:15.455] event: ReceiptReceived(DELIVERY from=+4***...528, count=1)
[20:10:15.521] receipt: DELIVERY from +4***...528 -> Delivered
[20:10:15.554] event: ReceiptReceived(DELIVERY from=+4***...858, count=1)
[20:10:15.556] receipt: DELIVERY from +4***...858 -> Delivered
[20:10:15.557] event: ReceiptReceived(DELIVERY from=+4***...910, count=1)
[20:10:15.559] event: ReceiptReceived(DELIVERY from=+4***...410, count=1)
[20:10:15.559] receipt: DELIVERY from +4***...910 -> Delivered
[20:10:15.559] receipt: DELIVERY from +4***...410 -> Delivered
[20:10:15.561] event: ReceiptReceived(READ from=+4***...858, count=1)
[20:10:15.564] receipt: READ from +4***...858 -> Read
[20:10:16.860] event: ReceiptReceived(DELIVERY from=+4***...747, count=1)
[20:10:16.862] receipt: DELIVERY from +4***...747 -> Delivered
[20:10:19.975] event: ReceiptReceived(READ from=+4***...910, count=1)
[20:10:20.009] receipt: READ from +4***...910 -> Read
```

--- a/src/app.rs
+++ b/src/app.rs
@@ -15,6 +15,13 @@ use crate::keybindings::{self, BindingMode, KeyAction, KeyBindings};
 use crate::theme::{self, Theme};
 use crate::signal::types::{Contact, Group, IdentityInfo, LinkPreview, Mention, MessageStatus, PollData, PollOption, PollVote, Reaction, SignalEvent, SignalMessage, StyleType, TextStyle, TrustLevel};
 
+/// Sentinel lifetime for paste temp files awaiting send confirmation from signal-cli.
+/// If signal-cli never confirms, the file is deleted after this many seconds.
+pub const PASTE_CLEANUP_SENTINEL_SECS: u64 = 3600;
+
+/// How long after send confirmation to wait before deleting a paste temp file.
+const PASTE_CLEANUP_DELAY_SECS: u64 = 10;
+
 /// Find the byte position one character forward from `pos` in `buf`.
 fn next_char_pos(buf: &str, pos: usize) -> usize {
     if pos >= buf.len() { return buf.len(); }
@@ -444,6 +451,10 @@ pub struct App {
     pub pending_attachment: Option<PathBuf>,
     /// Directory for temporary clipboard paste files (PID-scoped to avoid conflicts)
     pub paste_temp_path: PathBuf,
+    /// Paste temp files pending deletion: rpc_id → (path, delete_after)
+    /// Populated when a paste attachment send is dispatched; deletion deferred 10s after
+    /// signal-cli confirms or fails the send, to avoid deleting before signal-cli reads the file.
+    pub pending_paste_cleanups: HashMap<String, (PathBuf, Instant)>,
     /// Reply target: (author_phone, body_snippet, timestamp_ms)
     pub reply_target: Option<(String, String, i64)>,
     /// Delete confirmation overlay visible
@@ -2819,6 +2830,7 @@ impl App {
             file_browser_filtered: Vec::new(),
             file_browser_error: None,
             pending_attachment: None,
+            pending_paste_cleanups: HashMap::new(),
             paste_temp_path: {
                 let dir = std::env::temp_dir().join(format!("siggy-paste-{}", std::process::id()));
                 // Best-effort: clean any stale files from a previous run with the same PID,
@@ -5052,6 +5064,13 @@ impl App {
     }
 
     fn handle_send_timestamp(&mut self, rpc_id: &str, server_ts: i64) {
+        // Schedule any paste temp file for deletion after the delay (signal-cli has confirmed send)
+        if let Some((path, _)) = self.pending_paste_cleanups.remove(rpc_id) {
+            self.pending_paste_cleanups.insert(
+                rpc_id.to_string(),
+                (path, Instant::now() + std::time::Duration::from_secs(PASTE_CLEANUP_DELAY_SECS)),
+            );
+        }
         if let Some((conv_id, local_ts)) = self.pending_sends.remove(rpc_id) {
             crate::debug_log::logf(format_args!(
                 "send confirmed: conv={} local_ts={local_ts} server_ts={server_ts}",
@@ -5088,6 +5107,13 @@ impl App {
     }
 
     fn handle_send_failed(&mut self, rpc_id: &str) {
+        // Schedule any paste temp file for deletion after the delay (signal-cli has finished with it)
+        if let Some((path, _)) = self.pending_paste_cleanups.remove(rpc_id) {
+            self.pending_paste_cleanups.insert(
+                rpc_id.to_string(),
+                (path, Instant::now() + std::time::Duration::from_secs(PASTE_CLEANUP_DELAY_SECS)),
+            );
+        }
         if let Some((conv_id, local_ts)) = self.pending_sends.remove(rpc_id) {
             let mut found = false;
             if let Some(conv) = self.conversations.get_mut(&conv_id) {
@@ -6463,6 +6489,19 @@ impl App {
         }
     }
 
+    /// Delete any paste temp files whose 10s delay has elapsed.
+    /// Called each tick from the main event loop.
+    pub fn cleanup_paste_files(&mut self) {
+        self.pending_paste_cleanups.retain(|_rpc_id, (path, delete_after)| {
+            if Instant::now() >= *delete_after {
+                let _ = std::fs::remove_file(path);
+                false
+            } else {
+                true
+            }
+        });
+    }
+
     // --- Mouse support ---
 
     /// Returns true if any overlay is currently visible (mouse events should be ignored).
@@ -7658,6 +7697,81 @@ mod tests {
             app.conversations[conv_id].messages[0].status,
             Some(MessageStatus::Failed)
         );
+    }
+
+    // --- Paste cleanup tests ---
+
+    #[rstest]
+    fn send_timestamp_resets_paste_cleanup_deadline(mut app: App) {
+        // Set up a sentinel entry (far-future deadline = awaiting confirmation)
+        let tmp = std::env::temp_dir().join("test-paste-dummy.png");
+        let sentinel = Instant::now() + std::time::Duration::from_secs(PASTE_CLEANUP_SENTINEL_SECS);
+        app.pending_paste_cleanups.insert("rpc-1".to_string(), (tmp.clone(), sentinel));
+
+        app.handle_signal_event(SignalEvent::SendTimestamp {
+            rpc_id: "rpc-1".to_string(),
+            server_ts: 0,
+        });
+
+        // Deadline should now be ~10s from now, well under the sentinel
+        let (_, deadline) = app.pending_paste_cleanups.get("rpc-1").expect("entry should still exist");
+        let remaining = deadline.saturating_duration_since(Instant::now());
+        assert!(
+            remaining <= std::time::Duration::from_secs(PASTE_CLEANUP_DELAY_SECS),
+            "deadline should be reset to ~{PASTE_CLEANUP_DELAY_SECS}s, got {remaining:?}"
+        );
+    }
+
+    #[rstest]
+    fn send_failed_resets_paste_cleanup_deadline(mut app: App) {
+        let tmp = std::env::temp_dir().join("test-paste-dummy-fail.png");
+        let sentinel = Instant::now() + std::time::Duration::from_secs(PASTE_CLEANUP_SENTINEL_SECS);
+        app.pending_paste_cleanups.insert("rpc-2".to_string(), (tmp.clone(), sentinel));
+
+        app.handle_signal_event(SignalEvent::SendFailed {
+            rpc_id: "rpc-2".to_string(),
+        });
+
+        let (_, deadline) = app.pending_paste_cleanups.get("rpc-2").expect("entry should still exist");
+        let remaining = deadline.saturating_duration_since(Instant::now());
+        assert!(
+            remaining <= std::time::Duration::from_secs(PASTE_CLEANUP_DELAY_SECS),
+            "deadline should be reset to ~{PASTE_CLEANUP_DELAY_SECS}s, got {remaining:?}"
+        );
+    }
+
+    #[rstest]
+    fn cleanup_paste_files_removes_file_after_deadline(mut app: App) {
+        // Create a real temp file
+        let tmp = std::env::temp_dir().join(format!("test-paste-cleanup-{}.png", std::process::id()));
+        std::fs::write(&tmp, b"fake image data").expect("write temp file");
+        assert!(tmp.exists());
+
+        // Insert with a deadline already in the past
+        let past = Instant::now() - std::time::Duration::from_secs(1);
+        app.pending_paste_cleanups.insert("rpc-3".to_string(), (tmp.clone(), past));
+
+        app.cleanup_paste_files();
+
+        assert!(!tmp.exists(), "temp file should have been deleted");
+        assert!(app.pending_paste_cleanups.is_empty(), "entry should be removed");
+    }
+
+    #[rstest]
+    fn cleanup_paste_files_keeps_file_before_deadline(mut app: App) {
+        let tmp = std::env::temp_dir().join(format!("test-paste-keep-{}.png", std::process::id()));
+        std::fs::write(&tmp, b"fake image data").expect("write temp file");
+
+        // Insert with a future deadline
+        let future = Instant::now() + std::time::Duration::from_secs(60);
+        app.pending_paste_cleanups.insert("rpc-4".to_string(), (tmp.clone(), future));
+
+        app.cleanup_paste_files();
+
+        // File should still exist; clean it up manually
+        assert!(tmp.exists(), "temp file should not have been deleted yet");
+        let _ = std::fs::remove_file(&tmp);
+        assert!(!app.pending_paste_cleanups.is_empty(), "entry should still be present");
     }
 
     #[rstest]

--- a/src/main.rs
+++ b/src/main.rs
@@ -601,16 +601,28 @@ async fn dispatch_send(
                 Ok(rpc_id) => {
                     debug_log::logf(format_args!("send: to={} ts={local_ts_ms}", debug_log::mask_phone(&recipient)));
                     app.pending_sends
-                        .insert(rpc_id, (recipient.to_string(), local_ts_ms));
+                        .insert(rpc_id.clone(), (recipient.to_string(), local_ts_ms));
+                    // Register any paste temp file for deferred deletion. The actual delete is
+                    // triggered after send confirmation; this sentinel keeps it alive until then.
+                    // Only one paste attachment per send is expected; break after the first match.
+                    for path in &attachments {
+                        if path.starts_with(&app.paste_temp_path) {
+                            let sentinel = Instant::now()
+                                + Duration::from_secs(app::PASTE_CLEANUP_SENTINEL_SECS);
+                            app.pending_paste_cleanups
+                                .insert(rpc_id.clone(), (path.clone(), sentinel));
+                            break;
+                        }
+                    }
                 }
                 Err(e) => {
                     app.status_message = format!("send error: {e}");
-                }
-            }
-            // Clean up any temp paste files after signal-cli has read them
-            for path in &attachments {
-                if path.starts_with(&app.paste_temp_path) {
-                    let _ = std::fs::remove_file(path);
+                    // RPC failed to send — delete temp file immediately (signal-cli never saw it)
+                    for path in &attachments {
+                        if path.starts_with(&app.paste_temp_path) {
+                            let _ = std::fs::remove_file(path);
+                        }
+                    }
                 }
             }
         }
@@ -1052,6 +1064,9 @@ async fn run_app(
 
         // Auto-clear clipboard after timeout
         app.check_clipboard_clear();
+
+        // Delete paste temp files that have passed their 10s post-send delay
+        app.cleanup_paste_files();
 
         // Dynamic mouse capture toggle from settings
         if let Some(enabled) = app.pending_mouse_toggle.take() {

